File: services/analysis/riftlens/domain/replay_errors.py
from __future__ import annotations

from collections.abc import Mapping
from enum import StrEnum
from typing import Literal

ReplayErrorSeverity = Literal["fatal", "retryable", "informational"]


class ReplayErrorCode(StrEnum):
    """Closed taxonomy for native-replay and gameplay-source failures."""

    ROFL_MISSING = "ROFL_MISSING"
    ROFL_UNREADABLE = "ROFL_UNREADABLE"
    ROFL_INVALID = "ROFL_INVALID"
    ROFL_NOT_RECOGNISED = "ROFL_NOT_RECOGNISED"
    ROFL_METADATA_UNPARSED = "ROFL_METADATA_UNPARSED"
    MATCH_ID_UNRESOLVED = "MATCH_ID_UNRESOLVED"
    MATCH_NOT_INGESTED = "MATCH_NOT_INGESTED"
    INSTALL_NOT_FOUND = "INSTALL_NOT_FOUND"
    INSTALL_INVALID = "INSTALL_INVALID"
    PATCH_INCOMPATIBLE = "PATCH_INCOMPATIBLE"
    REPLAY_API_DISABLED = "REPLAY_API_DISABLED"
    LIVE_GAME_IN_PROGRESS = "LIVE_GAME_IN_PROGRESS"
    LAUNCH_FAILED = "LAUNCH_FAILED"
    LAUNCH_REJECTED = "LAUNCH_REJECTED"
    LAUNCH_TIMEOUT = "LAUNCH_TIMEOUT"
    PLAYBACK_NOT_STARTED = "PLAYBACK_NOT_STARTED"
    PLAYBACK_NOT_ADVANCING = "PLAYBACK_NOT_ADVANCING"
    REPLAY_API_UNAVAILABLE = "REPLAY_API_UNAVAILABLE"
    REPLAY_API_TLS = "REPLAY_API_TLS"
    PLAYBACK_UNREADABLE = "PLAYBACK_UNREADABLE"
    PAUSE_FAILED = "PAUSE_FAILED"
    RESUME_FAILED = "RESUME_FAILED"
    SEEK_FAILED = "SEEK_FAILED"
    SEEK_OUT_OF_BOUNDS = "SEEK_OUT_OF_BOUNDS"
    SEEK_TOLERANCE_EXCEEDED = "SEEK_TOLERANCE_EXCEEDED"
    CAPABILITY_UNSUPPORTED = "CAPABILITY_UNSUPPORTED"
    CLOCK_UNMAPPED = "CLOCK_UNMAPPED"
    CLOCK_OUT_OF_BOUNDS = "CLOCK_OUT_OF_BOUNDS"
    CLOCK_CALIBRATION_FAILED = "CLOCK_CALIBRATION_FAILED"
    LIVE_CLIENT_DATA_UNAVAILABLE = "LIVE_CLIENT_DATA_UNAVAILABLE"
    ACTIVE_PLAYER_UNAVAILABLE = "ACTIVE_PLAYER_UNAVAILABLE"
    PLATFORM_UNSUPPORTED = "PLATFORM_UNSUPPORTED"
    SOURCE_NOT_READY = "SOURCE_NOT_READY"
    SESSION_LOST = "SESSION_LOST"


_DEFAULT_MESSAGE: dict[ReplayErrorCode, str] = {
    ReplayErrorCode.ROFL_MISSING: "Replay file is missing.",
    ReplayErrorCode.ROFL_UNREADABLE: "Replay file exists but could not be read.",
    ReplayErrorCode.ROFL_INVALID: "Replay file is not a valid .rofl.",
    ReplayErrorCode.ROFL_NOT_RECOGNISED: "This file doesn't look like a League replay.",
    ReplayErrorCode.ROFL_METADATA_UNPARSED: (
        "Replay metadata could not be parsed; patch and length are unknown."
    ),
    ReplayErrorCode.MATCH_ID_UNRESOLVED: (
        "Could not determine which match this replay belongs to."
    ),
    ReplayErrorCode.MATCH_NOT_INGESTED: (
        "This replay's match is not in RiftLens yet. Ingest the match first."
    ),
    ReplayErrorCode.INSTALL_NOT_FOUND: "League of Legends installation was not found.",
    ReplayErrorCode.INSTALL_INVALID: "League of Legends installation is incomplete.",
    ReplayErrorCode.PATCH_INCOMPATIBLE: "Replay patch does not match the installed client.",
    ReplayErrorCode.REPLAY_API_DISABLED: "Replay API is not enabled in the game client.",
    ReplayErrorCode.LIVE_GAME_IN_PROGRESS: (
        "A live League game or queue is active; replay launch was refused."
    ),
    ReplayErrorCode.LAUNCH_FAILED: "Failed to launch the replay in the game client.",
    ReplayErrorCode.LAUNCH_REJECTED: "The game client refused or exited immediately after launch.",
    ReplayErrorCode.LAUNCH_TIMEOUT: "Replay API did not become ready before the launch timeout.",
    ReplayErrorCode.PLAYBACK_NOT_STARTED: "Replay process started but playback has not begun.",
    ReplayErrorCode.PLAYBACK_NOT_ADVANCING: "Replay is unpaused but game time is not advancing.",
    ReplayErrorCode.REPLAY_API_UNAVAILABLE: "Local Replay API is not reachable.",
    ReplayErrorCode.REPLAY_API_TLS: "Replay API TLS verification failed.",
    ReplayErrorCode.PLAYBACK_UNREADABLE: "Playback time/state could not be read.",
    ReplayErrorCode.PAUSE_FAILED: "Replay pause did not hold game time.",
    ReplayErrorCode.RESUME_FAILED: "Replay resume did not advance game time.",
    ReplayErrorCode.SEEK_FAILED: "Replay seek did not change playback time.",
    ReplayErrorCode.SEEK_OUT_OF_BOUNDS: "Requested seek time is outside the source clock.",
    ReplayErrorCode.SEEK_TOLERANCE_EXCEEDED: "Seek landed outside the allowed tolerance.",
    ReplayErrorCode.CAPABILITY_UNSUPPORTED: "Gameplay source does not expose this capability.",
    ReplayErrorCode.CLOCK_UNMAPPED: "No clock mapping exists for this timestamp.",
    ReplayErrorCode.CLOCK_OUT_OF_BOUNDS: "Timestamp is outside the clock map bounds.",
    ReplayErrorCode.CLOCK_CALIBRATION_FAILED: (
        "Automatic replay clock calibration failed; using an estimated or manual map."
    ),
    ReplayErrorCode.LIVE_CLIENT_DATA_UNAVAILABLE: (
        "Live Client Data is unavailable during this replay (non-fatal)."
    ),
    ReplayErrorCode.ACTIVE_PLAYER_UNAVAILABLE: (
        "activeplayer is unavailable during this replay (non-fatal)."
    ),
    ReplayErrorCode.PLATFORM_UNSUPPORTED: "This gameplay source is not supported on this platform.",
    ReplayErrorCode.SOURCE_NOT_READY: "Gameplay source is not ready.",
    ReplayErrorCode.SESSION_LOST: "Replay window was closed. Reopen the replay to continue.",
}

_SEVERITY: dict[ReplayErrorCode, ReplayErrorSeverity] = {
    ReplayErrorCode.ROFL_MISSING: "fatal",
    ReplayErrorCode.ROFL_UNREADABLE: "fatal",
    ReplayErrorCode.ROFL_INVALID: "fatal",
    ReplayErrorCode.ROFL_NOT_RECOGNISED: "fatal",
    ReplayErrorCode.ROFL_METADATA_UNPARSED: "informational",
    ReplayErrorCode.MATCH_ID_UNRESOLVED: "retryable",
    ReplayErrorCode.MATCH_NOT_INGESTED: "retryable",
    ReplayErrorCode.INSTALL_NOT_FOUND: "fatal",
    ReplayErrorCode.INSTALL_INVALID: "fatal",
    ReplayErrorCode.PATCH_INCOMPATIBLE: "fatal",
    ReplayErrorCode.REPLAY_API_DISABLED: "fatal",
    ReplayErrorCode.LIVE_GAME_IN_PROGRESS: "retryable",
    ReplayErrorCode.LAUNCH_FAILED: "fatal",
    ReplayErrorCode.LAUNCH_REJECTED: "retryable",
    ReplayErrorCode.LAUNCH_TIMEOUT: "retryable",
    ReplayErrorCode.PLAYBACK_NOT_STARTED: "retryable",
    ReplayErrorCode.PLAYBACK_NOT_ADVANCING: "retryable",
    ReplayErrorCode.REPLAY_API_UNAVAILABLE: "retryable",
    ReplayErrorCode.REPLAY_API_TLS: "fatal",
    ReplayErrorCode.PLAYBACK_UNREADABLE: "retryable",
    ReplayErrorCode.PAUSE_FAILED: "fatal",
    ReplayErrorCode.RESUME_FAILED: "fatal",
    ReplayErrorCode.SEEK_FAILED: "fatal",
    ReplayErrorCode.SEEK_OUT_OF_BOUNDS: "fatal",
    ReplayErrorCode.SEEK_TOLERANCE_EXCEEDED: "fatal",
    ReplayErrorCode.CAPABILITY_UNSUPPORTED: "fatal",
    ReplayErrorCode.CLOCK_UNMAPPED: "fatal",
    ReplayErrorCode.CLOCK_OUT_OF_BOUNDS: "fatal",
    ReplayErrorCode.CLOCK_CALIBRATION_FAILED: "informational",
    ReplayErrorCode.LIVE_CLIENT_DATA_UNAVAILABLE: "informational",
    ReplayErrorCode.ACTIVE_PLAYER_UNAVAILABLE: "informational",
    ReplayErrorCode.PLATFORM_UNSUPPORTED: "fatal",
    ReplayErrorCode.SOURCE_NOT_READY: "retryable",
    ReplayErrorCode.SESSION_LOST: "retryable",
}


def replay_error_message(code: ReplayErrorCode) -> str:
    """Return the canonical message for ``code``. Assumes ``code`` is a known member."""
    return _DEFAULT_MESSAGE[code]


def replay_error_severity(code: ReplayErrorCode) -> ReplayErrorSeverity:
    """Return fatal/retryable/informational for ``code``. Assumes taxonomy is complete."""
    return _SEVERITY[code]


class ReplayError(Exception):
    """Typed failure for gameplay-source / native-replay operations. No HTTP assumed."""

    def __init__(
        self,
        code: ReplayErrorCode,
        message: str | None = None,
        *,
        details: Mapping[str, object] | None = None,
    ) -> None:
        self.code = code
        self.details = dict(details or {})
        super().__init__(message if message is not None else replay_error_message(code))

    @property
    def severity(self) -> ReplayErrorSeverity:
        """Return the taxonomy severity. Assumes ``code`` is a known member."""
        return replay_error_severity(self.code)

    def is_fatal(self) -> bool:
        """Return True when this error should abort a native-replay proof."""
        return self.severity == "fatal"

    def is_retryable(self) -> bool:
        """Return True when a later poll/retry might succeed."""
        return self.severity == "retryable"

    def is_informational(self) -> bool:
        """Return True when the error must not fail Replay API control proof."""
        return self.severity == "informational"

File: apps/desktop/src/renderer/features/review/replayErrors.ts
export type ReplayActionId =
  | 'choose_file'
  | 'retry'
  | 'find_league_install'
  | 'enable_replay_api'
  | 'attach_video'
  | 'continue_without_replay'
  | 'reopen_replay'
  | 'open_replay'
  | 'try_anyway'
  | 'ingest_match'
  | 'open_replay_match'
  | 'sign_in_api_key'
  | 'choose_participant'
  | 'retry_after_live_game'
  | 'pick_match'

export type ReplayErrorView = {
  code: string
  message: string
  actionId: ReplayActionId
  actionLabel: string
}

const ACTION_LABELS: Record<ReplayActionId, string> = {
  choose_file: 'Choose another file',
  retry: 'Retry',
  find_league_install: 'Find League install',
  enable_replay_api: 'Enable Replay API',
  attach_video: 'Attach video instead',
  continue_without_replay: 'Continue without replay',
  reopen_replay: 'Reopen replay',
  open_replay: 'Open Replay',
  try_anyway: 'Try anyway',
  ingest_match: 'Ingest this match',
  open_replay_match: 'Open its review?',
  sign_in_api_key: 'Sign in with API key',
  choose_participant: 'Choose participant',
  retry_after_live_game: 'Retry after the live game',
  pick_match: 'Choose another file'
}

const CODE_DEFAULTS: Record<string, { message: string; actionId: ReplayActionId }> = {
  PLATFORM_UNSUPPORTED: {
    message: 'League replay import is available on Windows and macOS.',
    actionId: 'attach_video'
  },
  ROFL_UNREADABLE: {
    message: 'Replay file exists but could not be read.',
    actionId: 'choose_file'
  },
  ROFL_NOT_RECOGNISED: {
    message: "This file doesn't look like a League replay.",
    actionId: 'choose_file'
  },
  ROFL_INVALID: {
    message: 'Replay file is not a valid .rofl.',
    actionId: 'choose_file'
  },
  ROFL_MISSING: {
    message: 'Replay file is missing.',
    actionId: 'choose_file'
  },
  MATCH_ID_UNRESOLVED: {
    message: 'Could not determine which match this replay belongs to.',
    actionId: 'choose_file'
  },
  MATCH_NOT_INGESTED: {
    message: "This replay's match is not in RiftLens yet. Ingest the match first.",
    actionId: 'ingest_match'
  },
  MATCH_IDENTITY_MISMATCH: {
    message: 'This replay belongs to a different match. Open its review?',
    actionId: 'open_replay_match'
  },
  RIOT_CREDENTIAL_MISSING: {
    message: 'A Riot API key is required to download this match. Sign in with a developer key first.',
    actionId: 'sign_in_api_key'
  },
  PARTICIPANT_REQUIRED: {
    message: 'Choose which participant this review should coach.',
    actionId: 'choose_participant'
  },
  INSTALL_NOT_FOUND: {
    message: 'League of Legends installation was not found.',
    actionId: 'find_league_install'
  },
  INSTALL_INVALID: {
    message: 'League of Legends installation is incomplete.',
    actionId: 'find_league_install'
  },
  REPLAY_API_DISABLED: {
    message: 'Replay API is not enabled in the game client.',
    actionId: 'enable_replay_api'
  },
  REPLAY_API_UNAVAILABLE: {
    message: 'Local Replay API is not reachable.',
    actionId: 'retry'
  },
  LIVE_GAME_IN_PROGRESS: {
    message: 'A live League game or queue is active; replay launch was refused.',
    actionId: 'retry_after_live_game'
  },
  PATCH_INCOMPATIBLE: {
    message: 'Replay patch does not match the installed client.',
    actionId: 'try_anyway'
  },
  LAUNCH_FAILED: {
    message: 'Failed to launch the replay in the game client.',
    actionId: 'retry'
  },
  SESSION_LOST: {
    message: 'Replay window was closed. Reopen the replay to continue.',
    actionId: 'reopen_replay'
  },
  CLOCK_CALIBRATION_FAILED: {
    message: 'Automatic replay clock calibration failed; using an estimated or manual map.',
    actionId: 'continue_without_replay'
  },
  SEEK_FAILED: {
    message: 'Replay seek did not change playback time.',
    actionId: 'retry'
  },
  SOURCE_NOT_READY: {
    message: 'Gameplay source is not ready.',
    actionId: 'open_replay'
  },
  SIDECAR_UNAVAILABLE: {
    message: 'RiftLens analysis sidecar is not ready.',
    actionId: 'retry'
  }
}

const ACTION_ALIASES: Record<string, ReplayActionId> = {
  choose_file: 'choose_file',
  retry: 'retry',
  find_league_install: 'find_league_install',
  browse_to_install: 'find_league_install',
  enable_replay_api: 'enable_replay_api',
  enable_replay_api_and_restart_client: 'enable_replay_api',
  attach_video: 'attach_video',
  continue_without_replay: 'continue_without_replay',
  reopen_replay: 'reopen_replay',
  open_replay: 'open_replay',
  try_anyway: 'try_anyway',
  ingest_match: 'ingest_match',
  open_replay_match: 'open_replay_match',
  sign_in_api_key: 'sign_in_api_key',
  choose_participant: 'choose_participant',
  retry_after_live_game: 'retry_after_live_game',
  pick_match: 'pick_match',
  ask_user_to_pick_match: 'pick_match'
}

export function replayErrorView(input: {
  code: string
  message?: string | null
  suggestedAction?: string | null
}): ReplayErrorView {
  const defaults = CODE_DEFAULTS[input.code]
  const actionId = resolveAction(input.suggestedAction, defaults?.actionId ?? 'retry')
  const message = (input.message && input.message.trim()) || defaults?.message || knownFallback(input.code)
  return {
    code: input.code,
    message,
    actionId,
    actionLabel: ACTION_LABELS[actionId]
  }
}

export function knownReplayErrorCodes(): string[] {
  return Object.keys(CODE_DEFAULTS)
}

function resolveAction(suggested: string | null | undefined, fallback: ReplayActionId): ReplayActionId {
  if (suggested && ACTION_ALIASES[suggested]) {
    return ACTION_ALIASES[suggested]
  }
  return fallback
}

function knownFallback(code: string): string {
  return `${code.replaceAll('_', ' ').toLowerCase()}.`
}
